### PR TITLE
fix(curriculum): Improve Sudoku solver

### LINKED ISSUE #52585: [python]: learn-classes-and-objects-by-building-a-sudoku-solver
**TO FIX POST BETA RELEASE**

**General Suggestions**

- ~Have Campers create/instantiate a class and call the `str`/print method so they can see some output in the early lessons~
- ~NO `we`, `us` -> use `you`~
- ~Test hints (text) should end in a period (`.`)~

**Fixes**

- ~3: when referencing "Board", use codeblocks~
- ~4: add indentation to seed code, and include class + method in editable region~
- ~4: is it an _"attribute"_ or a _"property"_ ? it's an attribute.~
- ~5: add indentation to seed~
- ~8: add empty line to seed~
- ~9: add empty line to seed~
- ~10: add indentation to seed~
- ~12: indentation inconsistent~
- ~13: instructions unclear about requirement, and use of `start`~
- ~14: add empty line to seed~
- ~14: break into separate steps with individual variables for each op~
- ~15: probably remove step as part of fixing 14~
- ~16: add empty line to seed~
- ~17: add empty line to seed~
- ~18: add empty line to seed~
- ~19: add empty line to seed~
- ~20: add indentation to seed~
- ~21: add empty line to seed~
- ~21: switch description and instruction - always try to place the instruction (actionable thing for Camper) at the end~
- ~22: add indentation to seed~
- ~22: switch description and instruction~
- ~23: add empty line to seed~
- ~25: switch description and instruction~
- ~25: unclear what "inner condition" refers to~
- ~28: add indentation to seed~
- ~28: told to include `pass`, but next lesson seed does not~
- ~29: add indentation to seed~
- ~30: add indentation to seed~
- ~31: add indentation to seed + include `try` in editable region~
- ~32: "the code"? explain what returns~
- ~33: add indentation + include `try` in editable region~
- ~34: add indentation + include `try` in editable region~
- ~36: add indentation to seed~
- ~39: add indentation to seed~
- ~40: add indentation to seed~

### PR DESCRIPTION
**Pull request by zairahira**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #52585

### CODE REVIEW

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-classes-and-objects-by-building-a-sudoku-solver/6596e147789804474b4efa47.md [hunk 1]
```diff
@@ -1,23 +1,46 @@
...
+  test: () => {
+    const transformedCode = code.replace(/\r/g, '');
+    const if_check = 'if next_empty is None:';
+    const if_block = __helpers.python.getBlock(transformedCode, if_check);
+    const { block_body, block_indentation, block_condition } = if_block;
+    console.log('if block', if_block);
+    console.log('Block body', block_body);
+    console.assert(block_body.match(/return\s+True/));
+  }
+})
```
[the remark this answers is not in the document]
Comment (Dario-DC) [reply curriculum/challenges/english/07-scientific-computing-with-python/learn-classes-and-objects-by-building-a-sudoku-solver/6596e147789804474b4efa47.md:43]: This works for me.

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-classes-and-objects-by-building-a-sudoku-solver/6596e2744c80694d2e9a9712.md [hunk 2]
```diff
@@ -1,34 +1,50 @@
...
+        const transformedCode = code.replace(/\r/g, "");
+        const if_check = "if self.is_valid(next_empty, guess):"
+        const if_block = __helpers.python.getBlock(transformedCode, if_check);
+        const { block_body, block_indentation, block_condition } = if_block;
+        console.log("if block", if_block);
+        console.log("Block body", block_body);
+        console.assert(block_body.match(/if\s+self\.solver\(\):/));
+    
+    }
+})
```
Comment (Dario-DC) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-classes-and-objects-by-building-a-sudoku-solver/6596e2744c80694d2e9a9712.md:29]:
```suggestion
({
    test: () => {
        const transformedCode = code.replace(/\r/g, "");
        const if_check = "if self.is_valid(next_empty, guess)"
        const if_block = __helpers.python.getBlock(transformedCode, if_check);
        const { block_body} = if_block;        
        assert(block_body.match(/if\s+self\.solver\(\):/));    
    }
})
```

Comment (Dario-DC) [reply curriculum/challenges/english/07-scientific-computing-with-python/learn-classes-and-objects-by-building-a-sudoku-solver/6596e2744c80694d2e9a9712.md:29]: Removed unused variables

File: curriculum/challenges/english/07-scientific-computing-with-python/learn-classes-and-objects-by-building-a-sudoku-solver/6596e2744c80694d2e9a9712.md [hunk 3]
```diff
@@ -1,34 +1,50 @@
...
+        const transformedCode = code.replace(/\r/g, "");
+        const if_check = "if self.solver():"
+        const if_block = __helpers.python.getBlock(transformedCode, if_check);
+        const { block_body, block_indentation, block_condition } = if_block;
+        console.log("if block", if_block);
+        console.log("Block body", block_body);
+        console.assert(block_body.match(/return\s+True/));
+    
+    }
+})
```
Comment (Dario-DC) [inline curriculum/challenges/english/07-scientific-computing-with-python/learn-classes-and-objects-by-building-a-sudoku-solver/6596e2744c80694d2e9a9712.md:47]:
```suggestion
({
    test: () => {
        const transformedCode = code.replace(/\r/g, "");
        const if_check = "if self.solver()"
        const if_block = __helpers.python.getBlock(transformedCode, if_check);
        const { block_body} = if_block;        
        assert(block_body.match(/return\s+True/));    
    }
})
```

**Comment by moT01** [comment]
There's some failing tests here @zairahira. Do you want to take a look at those?

**Review by naomi-lgbt — APPROVED** [review]

**Review by majestic-owl448 — CHANGES_REQUESTED** [review]
I think it's missing continuity in the seed code, like for example, step 31 has the sudoku board in the seed code, the instructions don't say anything about removing it, and it's not present in step 32. Is that for some particular reason? Also, is there a particular reason the sudoku board is not present in the code since the beginning of the project? How can we build the project without having data to test it with?

**Review by Dario-DC — COMMENTED** [review]
I see two main issues with this project:
- the tests allow you to write the code with wrong indentation and pass (I know this was not the purpose of this PR)
- the project is almost built from the top to the bottom of the prototype and having the part that prints the board at the beginning it's not ideal.

My suggestion is to restructure the project to have the example board in the seed code from the beginning, give more information about objects, instances and methods before creating the __init__ method (adding brief examples not strictly related to the project). The construction of the board should occur at the end of the project.

**Pull request closed without merging**